[Liu Muchen] iP

LiuMC-SG opened the pull request:
### Duke
> "“Your mind is for having ideas, not holding them.” – David Allen ([Source](https://dansilvestre.com/productivity-quotes))

Duke is a chatbot that frees your mind of having to remember the things you need to do. It's

- Text-based
- Easy to learn
- Simple
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. Download it from [here](https://github.com/LiuMC-SG/ip/releases/tag/A-Jar)
2. Double-click the .jar file
3. Add all of your tasks
4. Let it manage those tasks for you :D

And it is **FREE** 🤡 ~~(*Totally* not as all the other dukes and replicas done from this [repo](https://github.com/nus-cs2103-AY2223S2/ip))~~

Features:

- [X] Managing tasks
- [X] Managing deadlines
- [ ] Reminders
- [X] GUI interface
- [ ] Being good at the game

---

If you are a random programmer that's looking through this, you can use it to practice Java as well. Here is the `main` code for the launcher.

```java
public static void main(String[] args) {
    Application.launch(Main.class, args);
}
```

Commit:
Add Gradle support

LiuMC-SG committed:
(Update): Skeletal version of Duke

LiuMC-SG committed:
(Update): Added ability to store user text and display when requested

LiuMC-SG committed:
(Update): Added tasks

LiuMC-SG committed:
(Update): Ability to mark and unmark tasks

Changed String to Task for more versatility
Switched to cases for convenience

LiuMC-SG committed:
(Update): Added ToDos, Deadlines and Events

LiuMC-SG committed:
(Update): Semi-auto testing using input/output comparison

LiuMC-SG committed:
(Update): Added error prompts for commands

LiuMC-SG committed:
(Update): Ability to delete tasks

LiuMC-SG committed:
(Update): Removed all commands syntax outside to another class

LiuMC-SG committed:
(Update): Add ability to save and load data

Tasks added into the list are able to be saved on loaded on each startup

LiuMC-SG committed:
Merge branch 'branch-Level-7'

LiuMC-SG committed:
(Update): Changed date storage to DateTime instead of String

LiuMC-SG committed:
(Fix): Spaces in empty lines removed

LiuMC-SG committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/DukeException.java
#	src/main/java/Event.java

Require fixing on toString for Deadline and Event

LiuMC-SG committed:
(Fix): Resolving errors arising from merge conflicts

Removed additional content
Ensured datetime worked correctly

LiuMC-SG committed:
(Fix): Format of documents with auto formatter

LiuMC-SG committed:
(Fix): Added marked/done status (missing previously)

LiuMC-SG committed:
(Fix): Refractor of commands to seperate classses

Ui: Deals with interactions with the user
Storage: Deals with loading tasks from the file and saving tasks in the file
Parser: Deals with making sense of the user command
TaskList: Contains the task list (e.g., it has operations to add/delete tasks in the list)

LiuMC-SG committed:
(Update): Allocated classes into respective packages

LiuMC-SG committed:
Merge remote-tracking branch 'origin/add-gradle-support'

LiuMC-SG committed:
(Update): Added JUnit test for tasks and its subclasses

LiuMC-SG committed:
(Fix): Auto format and optimisation

LiuMC-SG committed:
(Fix): Changed to proper main class

LiuMC-SG committed:
(Update): Fix Coding Standard

LiuMC-SG committed:
(Fix): Add name to JUnit tests

LiuMC-SG committed:
(Update): Add find command

Users can find description with matching substring of the given
input

LiuMC-SG committed:
Merge branch 'branch-Level-9'

LiuMC-SG committed:
(Update): Add JavaDoc for command classes

LiuMC-SG committed:
(Update): Add remaining JavaDocs

Fixed spacing for the previous command JavaDocs

LiuMC-SG committed:
Merge branch 'branch-A-JavaDoc'

LiuMC-SG committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/DukeException.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/command/AddDeadlineCommand.java
#	src/main/java/duke/command/AddEventCommand.java
#	src/main/java/duke/command/AddTodoCommand.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/ExitCommand.java
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/command/MarkCommand.java
#	src/main/java/duke/command/UnmarkCommand.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java

LiuMC-SG committed:
(Fix): Fix errors arising from merge conflict

LiuMC-SG committed:
(Fix): Fixed JavaDoc and Added for FindCommand

Jiayan-Lim commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,81 @@
+            try {
+                String fullCommand = UI.readCommand();
+                UI.showLine(); // show the divider line ("_______")
+                Command c = Parser.parse(fullCommand);
```
I like how you take advantage of abstraction to handle different kind of commands and make the code look nice and clean. 👍

LiuMC-SG replied:
Okay.

Jiayan-Lim commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,123 @@
+     * @return A corresponding task to the provided data
+     * @throws DukeException If the string representation does not match any format of tasks
+     */
+    private Task fromData(String data) throws DukeException {
```
Perhaps the method name can start with a verb, ie: readFromFile, so that it will comply with Java naming coding standard. I noticed the same issue in several other places too.

LiuMC-SG replied:
Except for fromData -> readFromData, would toData also count

Jiayan-Lim commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,49 @@
+    /**
+     * The finding term.
+     */
+    private final String TERM;
```
Perhaps can name it "KEYWORD" instead of "TERM".

LiuMC-SG replied:
Good suggestion. Thanks!

Jiayan-Lim commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * The due date of the deadline task.
+     */
+    protected LocalDate by;
```
Perhaps this instance attribute could be declared as private to better comply with Java information hiding.

LiuMC-SG replied:
Okay thanks. Realised I missed it in my event class too.

Jiayan-Lim commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,98 @@
+     * The list of commands that are available for use.
+     */
+
+    public enum AvailableCommands {
+        EXIT,
+        LIST,
+        MARK,
+        UNMARK,
+        DELETE,
+        ADD_TODO,
+        ADD_DEADLINE,
+        ADD_EVENT,
+        FIND
+    }
```
I like the way you use enum for different types of commands! 👍

LiuMC-SG replied:
Ok.

Jiayan-Lim reviewed (commented):
LGTM! Overall, the code is neat and well-structured! Just some naming issues need to fix.

LiuMC-SG committed:
(Update): Add checkstyle

LiuMC-SG committed:
(Fix): Fix checkstyle errors

LiuMC-SG committed:
(Update): Change CLI to GUI

LiuMC-SG committed:
Merge branch 'branch-Level-10'

LiuMC-SG committed:
(Fix): ConcatMsg uses join instead

LiuMC-SG committed:
(Fix): Fix protected to private for deadline, event class

LiuMC-SG committed:
(Fix): Change "TERM" to "KEYWORD" in find command

LiuMC-SG committed:
(Fix): Change method name to start with verb

All methods called 'fromData' changed to 'readFromData'

LiuMC-SG committed:
(Update): Add JavaDoc for new GUI classes

seanfirefox commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,123 @@
+    /**
+     * The file path where the tasks data are stored.
+     */
+    private final String FILE_PATH;
```
Should the naming be in this format? Consider changing to private static final.

LiuMC-SG replied:
I shall use filePath then since duke is defining the file path

seanfirefox commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,123 @@
+     * @param taskList The list of tasks to be saved
+     * @throws DukeException If there is an error saving the data of the tasks
+     */
+    public void updateData(TaskList taskList) throws DukeException {
```
Consider using plural form on collection of objects, for example ```TaskList tasks```

seanfirefox commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,58 @@
+    /**
+     * The list of tasks.
+     */
+    private final ArrayList<Task> TASKS;
```
Similar comment as above, perhaps your intention was to use private static final for this naming convention.

LiuMC-SG replied:
Okay although i shall need it to not be static to be able to initialise with pre existing tasks, so i shall change the naming convention

seanfirefox commented on src/main/java/duke/textui/TextUi.java:
```diff
@@ -0,0 +1,132 @@
+    /**
+     * A scanner that scans for user input.
+     */
+    private final Scanner SCANNER;
```
In this case consider using ```scanner``` instead of ```SCANNER``` if it is just a private final instead of private static final

LiuMC-SG replied:
Okay

seanfirefox reviewed (changes requested):
Most coding standard guidelines are adhered to. Good job!

LiuMC-SG committed:
Modify all files for code quality

All of the files had some form of code quality violation that needed
to be changed to fit with coding guidelines.

Let's change the errors in code quality.

Changing naming conventions of variables to match the camelcase
convention if they are not constant static variables for the class.

String representation of enums are moved into the enum itself for
simplicity when storing data. (i.e. AvailableCommands enum have
their command string representation within the enum now)

Duplicate code has been replaced by a function. (i.e. Checking if
the length of string list after splitting is 2 is abstracted out)

Long methods are broken up to relevant submethods.

Deleted dead code that passed on after switching over to GUI instead
of command line interface. (i.e. run() function in Duke was removed)

Magic string were changed to be a named constant instead were done
as well.

All of these changes would allow for better run-time efficienct,
robustness and understanding of the code.

restyled-commits committed:
Restyled by clang-format

LiuMC-SG committed:
Merge pull request #3 from LiuMC-SG/restyled/branch-A-CodeQuality

Restyle Modify all files for code quality

LiuMC-SG committed:
Merge pull request #2 from LiuMC-SG/branch-A-CodeQuality

Modify all files for code quality

LiuMC-SG committed:
Merge pull request #4 from LiuMC-SG/master

Updating branch from master

LiuMC-SG committed:
Add assert statements

Adding assertions allows us to be sure that the correct flow of
code is happening during runtime.

Let's add assertions to code statements that check whether the split
operation really resulted in split of string into 2. Moreover,
checking if resources are null for images.

All of these prevents exceptions and errors due to programmer error
to occur during runtime.

LiuMC-SG committed:
(Fix): Format error due to Restyled.io

LiuMC-SG committed:
(Fix): Bye closes GUI

Enabled bye command to close GUI properly.

Fixed indentation issue in command class.

LiuMC-SG committed:
Merge pull request #5 from LiuMC-SG/branch-A-Assertions

Add assert statements

LiuMC-SG committed:
(Update): Add grade.yml for Continuous Integration

LiuMC-SG committed:
(Update): Help command available

Added help command to display help of all of the commands available.

Menu items about and commands have been updated to actually show an
alert box to show about the chatbot and all the commands that are
available.

LiuMC-SG committed:
(Update): Add application icon

LiuMC-SG committed:
(Fix): Fix style error

Removed whitespace, order or static variables and added final to
alert box.

LiuMC-SG committed:
(Fix): Add javadoc for helpcommand to fix checkstyle

LiuMC-SG committed:
(Update): Add UI image into docs file

LiuMC-SG committed:
(Fix): Fix main class name for gradle

LiuMC-SG committed:
(Fix): JavaFX libraries fixed to have release work for linux

LiuMC-SG committed:
(Fix): Javadoc updated and fixed

LiuMC-SG committed:
(Update): Partial completion of README.md

LiuMC-SG committed:
(Fix): Change `to` to `from` in delete message

LiuMC-SG committed:
(Update): Completion of README.md for user guide

LiuMC-SG committed:
(Fix): Fix date in Ui.png

LiuMC-SG committed:
(Fix): Change capital error message to small letter for description

LiuMC-SG committed:
(Fix): Fix error when reading file not being shown

LiuMC-SG committed:
(Fix): Fix Javadoc formatting error.

LiuMC-SG committed:
(Fix): Fix numbering pattern

LiuMC-SG committed:
(Fix): Fix formatting for xml

LiuMC-SG committed:
(Fix): Fix markdown numbering
